File: apps/directory/test/version_compare_test.js
// Checks the JavaScript side of the version-comparison contract against the shared vectors.
//
// The same file is read by the C++ test (remote60_version_compare_test) and the Android client's
// Kotlin test. Three implementations, one set of expected answers -- if they drift apart, one of
// the three suites fails rather than the disagreement shipping quietly.
//
// Needs no server, so it runs first in run.js and costs nothing.
'use strict';

const fs = require('fs');
const path = require('path');
const { compareVersions } = require('../version_compare');

const vectorsPath = process.argv[2] ||
  path.join(__dirname, '..', '..', 'shared', 'version_compare_vectors.txt');

let failures = 0;
let checks = 0;

function check(name, cond, detail) {
  checks++;
  if (!cond) failures++;
  console.log(`${cond ? 'PASS' : 'FAIL'}  ${name}${detail ? '  ' + detail : ''}`);
}

/** Normalises to -1/0/1 so a sign convention change cannot pass unnoticed. */
function signOf(v) {
  return v < 0 ? -1 : (v > 0 ? 1 : 0);
}

let text;
try {
  text = fs.readFileSync(vectorsPath, 'utf8');
} catch (err) {
  console.log(`FAIL  could not read vectors file: ${vectorsPath}  ${err.message}`);
  process.exit(2);
}

let vectors = 0;
text.split('\n').forEach((raw, index) => {
  const line = raw.replace(/\r$/, '');
  if (!line || line.startsWith('#')) return;

  const parts = line.split('|');
  if (parts.length !== 3) {
    check(`vector line ${index + 1}`, false, `malformed: ${line}`);
    return;
  }
  const [left, right, expectText] = parts;
  const expect = Number(expectText);
  vectors++;

  const label = `"${left}" vs "${right}"`;
  const got = signOf(compareVersions(left, right));
  check(label, got === expect, `expected ${expect} got ${got}`);

  // Reversing the arguments must reverse the sign. Not listed in the vectors file because it has
  // to hold for every one of them; stating it once is stronger than listing pairs.
  const reversed = signOf(compareVersions(right, left));
  check(`${label} (antisymmetric)`, reversed === -expect, `expected ${-expect} got ${reversed}`);
});

// A vectors file that silently became empty would otherwise report a clean pass.
check('vectors file was not empty', vectors > 0, `${vectors} vectors`);

// Non-string input has to compare as older rather than throw: this reads a manifest field that
// arrives over the wire. Not in the vectors file because it is a JS-only concern.
check('undefined compares older than a version', signOf(compareVersions(undefined, '0.0.1')) === -1);
check('null compares equal to empty', signOf(compareVersions(null, '')) === 0);

console.log(`\n${failures === 0 ? 'RESULT: ALL PASS' : 'RESULT: FAILED'}` +
            `  (${checks} checks, ${failures} failed, ${vectors} vectors from ${vectorsPath})`);
process.exit(failures === 0 ? 0 : 1);

File: apps/native_poc/src/version_compare_test.cpp
// Checks the C++ side of the version-comparison contract against the shared vectors.
//
// The point of reading apps/shared/version_compare_vectors.txt rather than hard-coding cases here
// is that the same file is read by the directory server's JS test and the Android client's Kotlin
// test. Three implementations, one set of expected answers -- if they ever drift apart, one of the
// three suites fails rather than the disagreement shipping quietly.
//
// The vectors path comes from CMake (REMOTE60_VERSION_VECTORS_PATH) and can be overridden by
// argv[1], so the test can be pointed at a scratch file while working on it.

#include "version_compare.hpp"

#include <cstdio>
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>

#ifndef REMOTE60_VERSION_VECTORS_PATH
#define REMOTE60_VERSION_VECTORS_PATH "apps/shared/version_compare_vectors.txt"
#endif

namespace {

using remote60::native_poc::compare_versions;

int gFailures = 0;
int gChecks = 0;

/** Normalises to -1/0/1 so a sign convention change cannot pass unnoticed. */
int sign_of(int v) { return v < 0 ? -1 : (v > 0 ? 1 : 0); }

void check(const std::string& name, bool ok, const std::string& detail) {
  ++gChecks;
  if (!ok) ++gFailures;
  std::cout << (ok ? "PASS  " : "FAIL  ") << name;
  if (!detail.empty()) std::cout << "  " << detail;
  std::cout << "\n";
}

std::wstring widen(const std::string& s) {
  // Vectors are ASCII by construction (digits, dots, and a few suffix characters), so a byte-wise
  // widening is exact here and keeps the test free of locale behaviour.
  return std::wstring(s.begin(), s.end());
}

}  // namespace

int main(int argc, char** argv) {
  const std::string path = (argc > 1) ? argv[1] : REMOTE60_VERSION_VECTORS_PATH;
  std::ifstream file(path);
  if (!file) {
    std::cerr << "FAIL  could not open vectors file: " << path << "\n";
    return 2;
  }

  std::string line;
  int lineNo = 0;
  int vectors = 0;
  while (std::getline(file, line)) {
    ++lineNo;
    if (!line.empty() && line.back() == '\r') line.pop_back();  // vectors file may travel as CRLF
    if (line.empty() || line[0] == '#') continue;

    const size_t first = line.find('|');
    const size_t second = (first == std::string::npos) ? std::string::npos : line.find('|', first + 1);
    if (first == std::string::npos || second == std::string::npos) {
      check("vector line " + std::to_string(lineNo), false, "malformed: " + line);
      continue;
    }
    const std::string left = line.substr(0, first);
    const std::string right = line.substr(first + 1, second - first - 1);
    const std::string expectText = line.substr(second + 1);
    const int expect = std::stoi(expectText);
    ++vectors;

    std::ostringstream label;
    label << "\"" << left << "\" vs \"" << right << "\"";

    // Wide, which is what the installer actually passes.
    const int wide = sign_of(compare_versions(widen(left), widen(right)));
    check(label.str() + " (wide)", wide == expect,
          "expected " + std::to_string(expect) + " got " + std::to_string(wide));

    // Narrow, which is what a future updater reading a manifest would pass.
    const int narrow = sign_of(compare_versions(left, right));
    check(label.str() + " (narrow)", narrow == expect,
          "expected " + std::to_string(expect) + " got " + std::to_string(narrow));

    // Reversing the arguments must reverse the sign. This is not in the vectors file because it
    // has to hold for every one of them, and stating it once is stronger than listing pairs.
    const int reversed = sign_of(compare_versions(widen(right), widen(left)));
    check(label.str() + " (antisymmetric)", reversed == -expect,
          "expected " + std::to_string(-expect) + " got " + std::to_string(reversed));
  }

  // A vectors file that silently became empty would otherwise report a clean pass.
  check("vectors file was not empty", vectors > 0, std::to_string(vectors) + " vectors");

  std::cout << "\n" << (gFailures == 0 ? "RESULT: ALL PASS" : "RESULT: FAILED")
            << "  (" << gChecks << " checks, " << gFailures << " failed, "
            << vectors << " vectors from " << path << ")\n";
  return gFailures == 0 ? 0 : 1;
}
